### DISCUSSION TITLE
Question about indices from FastNS.get_pairs()

### DISCUSSION THREAD

**Discussion by hrwusht** [MDAnalysis/mdanalysis#4488] [Q&A]
Hi,

I'm working with MDAnalysis and have a question about the indices returned by the `FastNS.search` (`NSResults.get_pairs()`) method.

Here's the relevant code I'm using:

```python
import MDAnalysis as mda
from MDAnalysis.lib.nsgrid import FastNS

u = mda.Universe('protein.pdb', 'md_center.xtc')
protein_atoms = u.select_atoms("protein")

pair_counts = np.zeros((len(protein_atoms), len(protein_atoms)))

groupA = u.select_atoms('resid 1-20 and not type H')
groupB = u.select_atoms('resid 20-30 and not type H')

ns = FastNS(cutoff=4.8, coords=groupA.positions, box=u.dimensions)
ns.search(groupB.positions)
pairs = ns.get_pairs()

for i, j in pairs:
    # Is it right for j representing groupA and i representing groupB?
    pair_counts[groupA[j].index, groupB[i].index] += 1
```
My understanding is that i represents the index of an atom from groupB, and j represents the index of an atom from groupA. Is this correct?

Any help would be greatly appreciated! 

Thank you!
Shiauthie Wu

**Comment by orbeckst** [comment]
> My understanding is that i represents the index of an atom from groupB, and j represents the index of an atom from groupA. Is this correct?

Based on the docs https://docs.mdanalysis.org/stable/documentation_pages/lib/nsgrid.html#MDAnalysis.lib.nsgrid.FastNS.search that would be my interpretation.

@hmacdope @richardjgowers does this sound right?

**Reply by hrwusht** [reply]
I'm trying to understand the neighbor search functionality in MDAnalysis, and I noticed what appears to be an inconsistency in the naming conventions used for the indices i and j in the documentation.

In the documentation you provided, the indices i and j are described as follows:
1. From the `MDAnalysis.lib.nsgrid.FastNS.search` documentation:

> "The search is exclusive i.e. only the pairs (i, j) such that atom[i] from **query** atoms and atom[j] from the **initialized set of coordinates** is stored as neighbors."

_Here, i is referred to as representing the **query** atoms, and j is referred to as representing the **initialized set of coordinates**._

2. From the `MDAnalysis.lib.nsgrid.NSResults.get_pairs` documentation:

> "Returns an array of shape (N, 2), where N is the number of pairs between reference and configuration within the specified distance. For every pair (i, j), **reference**[i] and **configuration**[j] are atom positions such that reference is the position of query atoms while configuration contains the position of group of atoms used to search against the query atoms."

_In this case, i is referred to as representing the **reference** atoms, and j is referred to as representing the **configuration** atoms._

It seems inconsistent to use different terminology ("query" vs. "reference") to refer to the same groups when describing the indices i.

Any clarification would be greatly appreciated.

Thank you!

**Reply by orbeckst** [reply]
How would you rephrase the docs to make them clearer?

@hmacdope @richardjgowers do you have an opinion?

**Reply by hmacdope** [reply]
Sorry for the slow reply, so my understanding is that we use two sets of nomenclature, where we probably could just one

M understanding is that 

query[i]  == reference[i]
coordinates[j] == configuration[j]

If this is correct (@richardjgowers?), I think we should probably go with changing to `reference` and `configuration` IMO as these make the most logical sense (to me).



reference

**Reply by orbeckst** [reply]
@erwusht could you open an issue to improve the docs for FastNS? Following @hmacdope 's assessment, we should use consistent terminology. 

Please summarize the problem, suggest the change, and link to this discussion.

Once we have an issue, we know that this is something that should get fixed and then maybe you or another volunteer can make the changes. If it's not an issue then it will be forgotten.

Thank you!

**Reply by hrwusht** [reply]
@orbeckst @hmacdope  Sorry for the delayed response. 

Thank you very much for being willing to take the time to solve this problem. 

I'm willing to contribute towards this improvement of the docs. I'm also happy to offer some ideas as references for the rephrased versions. 

Maybe a few days later, I need some time to do this.

**Reply by orbeckst** [reply]
Any contributions are very welcome! I suggest starting with opening an issue because that's the point where we can then discuss what needs to be done. You can then just open a PR that references the issue.
